### ISSUE TITLE
investigate swapping out reqwest for hyper

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#362]
We are doing a whole lot of manual TLS work ourselves, along with painful re-exporting of feature sets which selects a lot of the same dependencies (which are further re-ordered by #360 where we get a second client for websockets).

It would be interesting to see how feasible it is to just run straight `hyper` within `kube` and see what the fallout of that would be.
What does this drop? What does this let us solve better?

I imagine this will force us to make a better encapsulation interface to the Client (to avoid impulses to go down #17), as we are now putting a bunch of parameters within it, making Clone more expensive, even more so after #360.

Most of the dependencies of reqwest we already have anyway; tokio, serde, various tls deps, pin, url, futures*, but [there is more](https://github.com/seanmonstar/reqwest/blob/1f425a0244bcd7b4565dceb9076450d951f2ec03/Cargo.toml#L71-L123). What extra dependencies would this mean we would have to lift into `kube`?

- async gzip compression, almost certainly - probably a good default to have
- cookies, can probably drop?
- dns resolver?
- error types: [`ReqwestError`](https://docs.rs/reqwest/0.10.10/reqwest/struct.Error.html) is re-exported in [kube::Error](https://github.com/clux/kube-rs/blob/master/kube/src/error.rs) due to its nice interface. But looks like [Hyper::Error](https://docs.rs/hyper/0.14.2/hyper/struct.Error.html) could be usable unless we want to unpack the errors out out of reqwest/hyper entirely.

If anyone wants to investigate and/or provide some PoC PRs, this would be appreciated.

**Comment by kazk** [comment]
- TLS (improvement by deduplication)
    - Code in #360 was mostly taken from `reqwest` and we should be able to share the setup between HTTPS and WSS
    - https://github.com/hyperium/hyper-tls
    - https://github.com/ctz/hyper-rustls
- Compression: https://github.com/Nemo157/async-compression
- Proxy: https://github.com/tafia/hyper-proxy
- OAuth2: https://github.com/dermesser/yup-oauth2 (improvement?)

**Comment by jbg** [comment]
One small request regarding TLS: please expose the `vendored` feature on hyper-tls, as it's the easiest way handle the openssl dependency when cross-compiling. At the moment we have to add a direct dep on `reqwest` with the `native-tls-vendored` feature in order to get this feature activated, as it's not passed through by `kube`.

**Comment by kazk** [comment]
@jbg Can you open a new issue for `vendored`? We should be able to add a feature for it later. Comments are easy to forget.
